### ISSUE TITLE
Author intent not working with forwarding includes

### ISSUE THREAD

**Issue by jru** [include-what-you-use/include-what-you-use#536]
IWYU incorrectly assumes that the definition of `Type` is not required in the following test because the *author* doesn't explictly opt out (by forward declaring it in the same file), and suggests a removal that makes it break:

```cpp
//file: intent-fwd.h
struct Type; 

//file: intent-decl.h
#include "intent-fwd.h"
Type decl(); 

//file: intent-type.h
struct Type {}; 

//file: intent.cc
#include "intent-decl.h"
#include "intent-type.h"

void func()
{
  decl(); //Full-use of Type in return
}
```
```
intent.cc should add these lines:

intent.cc should remove these lines:
- #include "intent-type.h"  // lines 3-3

The full include-list for intent.cc:
#include "intent-decl.h"  // for decl
---
```

This shouldn't happen. IWYU should suggest to include `intent-type.h` unless it can see that the author of `Type decl()` already does.  

The discussion about this started in #535 

Some possibly related items: #186, #175, #90

**Comment by kimgr** [comment]
@jru I've been consistently impressed by your contributions, and wanted to ask if you're interested in joining the IWYU project more permanently. Can you contact me at kim.grasman@gmail.com and we can discuss? Thanks!

**Comment by jru** [comment]
Yes, I am pretty convinced too that where the forward-decl itself is should not matter. 
As you pointed out, if there is no definition the code would fail anyway. And in case the caller does not need the complete type, the same goes for the declaration: there must one reachable from it somewhere.

I am sure there will be some border cases to consider, but in principle it sounds reasonable to simplify the rule that way.

**Comment by jru** [comment]
Do you mean it is not desirable because it suggest removing *\<iosfwd\>*? 
It looks like a very similar scenario to the one above. Or am I missing something?

I think what's also happening here is that IWYU would not unwrap references when looking for type decls (in *TypeToDeclAsWritten*). At some point, the func should probably be rewritten to dereference all types before matching on them.

**Comment by kimgr** [comment]
No, I mean because the author-intent rule would say that since a forward-decl for `std::ostream` is available earlier in the translation unit, callers are responsible for including. When in fact it's just a coincidence that the includer had pulled in `<iosfwd>` before `myheader.h`.

I'd like for the author-intent rule to be able to follow the include graph and say that a forward-declare of symbol `T` in any file included from the using header (`intent-decl.h` in your example) would signal that includers were responsible for including a full definition.

Currently the rule only allows forward-declaration in the exact same file.

Interesting note about `TypeToDeclAsWritten`. I wonder if typedefs are part of why? E.g. `typedef const MyType* MyTypePtr;`.

**Comment by jru** [comment]
Yes, I think I see what you mean now. 

So summarizing, deciding on the intent and following actions for a type T:

 - _From the author perspective ("intent-decl.h"):_ 

     If a forward declaration of T is enough and present before in the file, or in its direct (or transitive?) includes, then mark the author as not responsible for the definition of T, otherwise mark it as responsible for it.

 - _From the caller perspective ("intent.cc"):_

   If T is used in a way that is definition is not required: check if the author is responsible for it, and if it isn't mark the declaration of T as desired ("intent-fwd.h") for the usage (either by keeping the include that forward declares it, or by suggesting a forward declaration).

   If T is used in a way that is definition is required: check if the author is responsible *and* if it includes the full definition of T. If both are true, mark the author site ("intent-decl.h") as desired, otherwise mark the definition of T as desired ("intent-type.h").


Something like this should cover for all the cases we have discussed and still allow the author to opt out of including all of its defintions. What do you think?


> Interesting note about TypeToDeclAsWritten..

I think typedefs of pointers should be ok. The problem is if a type is directly a pointer or a reference.

**Comment by kimgr** [comment]
I can't quite follow your summary, there are a couple of places where I think you confuse the declaration of `T` with the declaration of `decl`:

> [T is forward-declare-used and author has forward-declared] If it is, mark the author site ("intent-decl.h") as desired, 

`intent-decl.h` should be unconditionally desired becase `decl` is used.
And if T is forward-declare-used in `intent.cc`, I think IWYU should make sure `intent.cc` has its own forward declaration (possibly via `intent-fwd.h` if already included, but maybe IWYU will flatten that at a later stage anyway).

> If T is used in a way that its definition is required: check if the author is responsible and if it includes the full definition of T. If both are true, mark the author site ("intent-decl.h") as desired

It should already desired because of `decl`. The question is if `T` should trigger another include for `intent-type.h` (if author is responsible) or be satisfied with the reexport from `intent-decl.h` (if author added a hint forward-decl).

I think this may be what you're already saying, I just needed to break it apart to understand it myself :)

I think the high-level logic in `CodeAuthorWantsJustAForwardDeclare` is sane, it's just that the "forward-declare-in-same-file" needs to be loosened a little.

Good point about immediate vs transitive includes, by the way, I hadn't thought about that. Off the top of my head, I would say any forward-decl available in the transitive closure of `intent-decl.h` should count. But this will need some experimentation on real code...

**Comment by jru** [comment]
> I can't quite follow your summary, there are a couple of places where I think you confuse the declaration of T with the declaration of decl:

I admit it is confusing. I refer to a T type, but then make references to my example. Actually, I should have simply used *Type* instead of *T*. I'll amend it.

Other than that, I believe I am not confusing things. Whatever happens with *Type* is independent of the *decl()* and that's how I am making the analysis. The same way IWYU would. 

You should read it as thinking about *Type* in isolation. In this specific case, the later analysis will find the *intent-decl.h* desired because of *decl()*, but that should be irrelevant during the analysis of *Type*.

I hope it makes more sense now :)

**Comment by kimgr** [comment]
It doesn't make sense to me to think about `Type` entirely isolated from `decl`, because it is only in the context of `decl` that `Type` becomes re-exported or not (depending on the presence of a hint forward-decl).

But with that proviso, your description makes sense :)
